### ISSUE TITLE
Significant notification message loss in production

### ISSUE THREAD

**Issue by hannes-ucsc** [HumanCellAtlas/data-store#2333]
While researching https://github.com/DataBiosphere/azul/issues/1212 I found some disconcerting evidence that the DSS notifier is dropping messages.

Every notification message goes into the first notification queue (queue 0). Every notification message that goes into queue 0 should come out of that queue 0 again. Invariably, there should be as many message receptions for queue 0 as there are message additions to that queue. The notifier code emits a log message right before it adds a message and after it receives a message. The number of log messages should be equal for both message types. It isn't.

<img width="1494" alt="Screen Shot 2019-08-08 at 9 52 38 PM" src="https://user-images.githubusercontent.com/5143256/62755260-1f69db00-ba28-11e9-89f5-8f309d0b27c6.png">

<img width="1497" alt="Screen Shot 2019-08-08 at 9 52 50 PM" src="https://user-images.githubusercontent.com/5143256/62755269-25f85280-ba28-11e9-8830-0fc2da03220c.png">

Note the total number of hits: 228,475 additions against 44,629 receptions. Also notice how stretched out over time the reception graph is. The rate of the message consumption is also surprisingly low: during the first 4 days or so it is just 500 messages per hour, that's less than 10 per minute. IIRC, when I wrote the notifier it was able to handle many more messages, but the consumption rate was configurable via the number of worker threads in the pool.

My theory is that SQS is simply expiring message (the default retention period is 4 days and I am not sure I wrote code to extend the retention period.

This CloudWatch metric underlines that:

<img width="1422" alt="Screen Shot 2019-08-08 at 10 17 31 PM" src="https://user-images.githubusercontent.com/5143256/62756015-0d3d6c00-ba2b-11e9-88ae-4c9e43b73edb.png">

Note how the message age is capped at ~ 346k seconds which is 4 days.

TL;DR It looks as if queue consumption rate and message retention period are ill configured in production, causing SQS to expire notification messages before they even get their first shot of being delivered to the subscriber.

**Comment by DailyDreaming** [comment]
We can increase the number of elasticsearch instances to accommodate this.  However, it might be costly.

I don't think the lambdas are the bottleneck.

I also could be misunderstanding how the infrastructure works with elasticsearch here, but our instance is a `m4.large`, which we have two of for `dev`/`integration`/`staging` and three of for `prod` (https://github.com/HumanCellAtlas/data-store/blob/master/infra/elasticsearch/main.tf#L90).  These cost 0.40 cents an hour, or $9.60/day.  Increasing by one more node adds one to each stage ($9.60 x 4 = about $40/day).  Over a month (about $40/day * 30), that'd be $1200/month.

We should not encounter this error if we do not have any bad subscriptions (in which case the queue is flooded with messages that can never be delivered and take forever to expire).  Measures to counter this include checking for bad subscriptions (100% failure over the past 24 hours I think is a decent metric), and limiting the subscriptions one user can create (already in place now).

That should solve the problem in theory until the `query-service` can replace our `elasticsearch` instance and solve this as well as current restrictions on `bundle`/`collection` size.

**Comment by hannes-ucsc** [comment]
> We can increase the number of elasticsearch instances to accommodate this. However, it might be costly.

This issue has nothing to do with ES. I thought we cleared this up last week when we discussed this in person. When the notifications end up on the the first notifier queue, the document was already indexed, and the percolate query was already run to determine the matching subscriptions.

> I don't think the lambdas are the bottleneck.

They are. I wrote the notifier and I did the research on this issue and I give evidence in the ticket description. This is my code. So at this point I would appreciate it if you gave some evidence for why you think I am wrong. I did my part.

> We should not encounter this error if we do not have any bad subscriptions 

I don't think that's accurate. Here's my evidence: The logs show that messages are consumed at a rate of 10 per minute, message retention is 4 days. Any activity that creates a queue length of more than 4 * 24 * 60 * 10 == 57600 messages risks having messages dropped. You just need an influx of bundles at rate of 10 bundles per min and two subscriptions to get there. Tabula muris has 100k bundles. That's an entirely realistic scenario.

Harvesting "bad" subscriptions is a good idea. But I might not agree with your definition of "bad". It's entirely possible that a subscriber is down for a few days. The DSS should not drop their subscriptions just because there is a bug in the notifier. 

As I said a the end of my OP, you can either increase consumption rate by increasing the rate of worker lambda invocations or the number of threads in each worker lambda. You can also prioritize the first queue by assigning more threads to it. Those measures would increase Lambda cost. You can also increase the message retention to account for the lower message consumption.

What frustrates me is that the solution is relatively simple and straight forward. It would take less time to implement than what I already spent on this ticket. It took me 2 hours to diagnose it. I went to your standup, I pinged your PM, I talked to you for 30min, to one of your team members for 30 min. So I easily spent 3.5 hours on it. My estimate for this ticket would be 2h of work.

A word on your reasoning, too: It's obviously contradictory. You first claim (without justification) that ES is the problem and that addressing it by adding nodes would be too expensive. But then you argue in terms of the queues, use terms like "queue flooding" and message expiry. It's either ES or the queue are flooded. It can't be both, because notifications are processed AFTER a bundle was indexed into ES.
